react-carousel/src/tools/carouselPluginResolver.js

```diff
@@ -0,0 +1,107 @@
+import _flatten from 'lodash/flatten';
+import _sortBy from 'lodash/sortBy';
+
+import { plugins as pluginsFunc } from '../constants/plugins';
+import pluginsOrder from '../constants/pluginsOrder';
+
+const carouselPluginResolver = (
+  plugins,
+  carouselProps,
+  trackContainerRef,
+  carouselRef,
+) => {
+  const carouselPlugins = carouselProps?.plugins.map((plugin) => {
+    if (typeof plugin === 'string') {
+      return (
+        pluginsFunc[plugin.toUpperCase()] &&
+        pluginsFunc[plugin.toUpperCase()]({
+          carouselProps: {
+            ...carouselProps,
+            children: carouselProps.children
+              ? carouselProps.children
+              : carouselProps.slides,
+          },
+          options: plugin.options,
+          refs: { trackContainerRef, carouselRef },
+        })
+      );
+    }
+    return plugin.resolve({
+      carouselProps,
+      options: plugin.options,
+      refs: { trackContainerRef, carouselRef },
+    });
+  });
+  const itemClassNames = _flatten(
+    carouselPlugins.map(
+      (plugin) =>
+        plugin.itemClassNames &&
+        plugin.itemClassNames({
+          carouselProps,
+          options: plugin.options,
+          refs: { trackContainerRef, carouselRef },
+        }),
+    ),
+  ).filter((className) => typeof className === 'string');
+
+  const carouselClassNames = _flatten(
+    carouselPlugins.map(
+      (plugin) =>
+        plugin.carouselClassNames &&
+        plugin.carouselClassNames({
+          carouselProps,
+          options: plugin.options,
+          refs: { trackContainerRef, carouselRef },
+        }),
+    ),
+  ).filter((className) => typeof className === 'string');
+
+  const carouselCustomProps = Object.assign(
```

Comment:
could you use everywhere object spread operator instead of `Object.assign`?

Reply:
done

Reply:
The result of applying the spread operator will be different in this case resulting in breaking the code

Reply:
could you explain why?

you don't modify any object so the following should do the same (btw. you have already an object spread operator):
```
{
    ...carouselPlugins.map(
      (plugin) => plugin.carouselCustomProps && plugin.carouselCustomProps(),
    ),
}
```

Reply:
Ok. Let me describe how it will work in the situation when we have 2 plugins and only one of them has defined carouselCustomProps. 

Current code
```
  const carouselCustomProps = Object.assign(
    {},
    ...carouselPlugins.map(
      (plugin) => plugin.carouselCustomProps && plugin.carouselCustomProps(),
    ),
  );
```
It will give you the following result
```
{ onMouseMove: ƒ }
```

And in your proposed way:
```
  const carouselCustomProps = {
    ...carouselPlugins.map(
      (plugin) => plugin.carouselCustomProps && plugin.carouselCustomProps(),
    ),
  };
```
the result will be:  
```
{
  0: {onMouseMove: ƒ}
  1: undefined
}
```

As you can see the undefined values aren't stripped also we need to destructure this object later. Besides, I think this changing Object.assign to spread isn't not an important issue in beta release

Reply:
it's not very important for me but still I cannot understand.
```
> x={foo:123,bar:456,baz:undefined}
{ foo: 123, bar: 456, baz: undefined }
> {...x}
{ foo: 123, bar: 456, baz: undefined }
> Object.assign({}, x)
{ foo: 123, bar: 456, baz: undefined }
```

Reply:
our case is a little bit different
```
> arr = [undefined, { f1: () => {}, f2: () => {} }];
[undefined, { f1: () => {}, f2: () => {} }]
> {...arr}
{0: undefined, 1: { f1: () => {}, f2: () => {} } }
>Object.assign({}, arr);
{ f1: () => {}, f2: () => {} }
```
as you can see in the second scenario the object is already flattened and without undefined values